ENH,DEP: Add DTypePromotionError and finalize the == and != FutureWarning/Deprecation

seberg opened the pull request:
Opening as a draft, because I think I should probably split it out the pieces (although it isn't as huge as one might think).

The PR effectively does a few things:
1. Add a new `DTypePromotionError` exception and use it for all promotion errors (DType or common instance where promotion shouldn't work).
2. Reorganize ufunc dispatching very mildly to chain this error when it occurs with a `UFuncNoLoopError`, also make the two-operand version inherit from the first.
3. Prohibit `np.equal(timedelta, datetime)`, etc. to work.  This is a very small break, because `np.equal(timedelta, datetime, casting="unsafe")` actually used to work.
4. Finalize the derprecation/futurewarning :).  When dtypes are fundamentally not comparable, an array of all False or all True is returned.

There are some tiny things to (possibly consider):
* Object arrays could cause a nested error.  That could probably be mitigated, but since `==` calls `==` in the ufunc loop NumPy arrays can't fall into that trap easily.
* The timedelta/datetime change is a "breaking change", I think its fine, it is just too absurd and ridiculously niche (with the need for `casting=`).

**The main changes to think about:**
* Should we push this into the ufunc?  This *could* be done.  It would require either some hacks for void, or properly moving void comparisons into the ufunc code (which is a larger project probably).  *I think this could also be done at a later time.*
* What errors do we want?  It might actually be sufficient to have a single reliable `DTypeTypeError` for everything.  I wouldn't mind introducing one as a baseclass for both `DTypePromotionError ` and the `UFunc*` errors (we can still delay exposing the `UFunc*` errors.  We could probably also get away without `DTypePromotionError `, but I think I like it, it is pretty nicely and narrowly defined.

seberg committed:
ENH: Add an InvalidPromotion exception

seberg committed:
ENH: Ensure UFuncTypeError is raised when promotion fails

If promotion fails internally, this effectively means that there is
no loop available for the given ufunc, so chain that error.

We only do this for `InvalidPromotion` since we assume other errors
may well be critical.

seberg committed:
DEP: Finalize the comparison FutureWarning/DeprecationWarning

This finalize the deprecation warning, there are a couple of corner
cases that we don't get as nicely as we could.  E.g. we (for now?)
manually create the all-false or all-true result array with a bit
clunky subclass support (subclasses can always override `==` and
`!=` though).
We also keep some (existing) behavior for 0-D objects where we
just return `NotImplemented`, which means the result should end
up with Python booleans (which is probably just as well).

seberg committed:
TST: Adapt test for modification to == and != deprecation/futurewarning

seberg committed:
MAINT: (equality) fixup subclass handling for new array-return path

seberg committed:
TST: Add test to cover basic subclass support of new code paths

seberg committed:
API: Always reject equal/not_equal for datetime/timedelta mix

This allows correctly cathing the error, also adjust the the NoLoop
error is used for the the "binary no loop error".
For now, I think I may want to keep the casting error distinct.

seberg committed:
TST: Remove old deprecation test and convert/add new ones

mhvk commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1074,33 +972,100 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+            Py_RETURN_NOTIMPLEMENTED;
+        }
+        /*
+         * Unfortunately, this path doesn't do the full __array_ufunc__
```
Does this matter? Didn't subclasses already get a chance when entering the `ufunc`?

seberg replied:
Ah, I hadn't thought of that, but I think it still does because of nested ufunc calls.  I.e. if Quantities would not implemented `==` but their `__array_ufunc__` calls `np.equal()` on base-class arrays the error (correctly) is floated up, but the result (incorrectly) would not be wrapped.

mhvk replied:
OK, I see, you mean something like `np.array([MyClass(...), MyClass(...)], dtype=object)`?

seberg replied:
I was thinking of something like:
```
class MyArr(np.ndarray):
    def __array_ufunc__(self, ufunc, ops, ...):
        array_ops = cast_self_to_numpy(ops)
        np_result = ufunc(array_ops, ...)
        return MyArr.from_numpy(np_result)

    def __eq__(self, other):
        return super().__eq__(self, other)  # gets called, but does nothing
```
where `==` does nothing and relies on `__array_ufunc__`.  In that case `ufunc(array_ops, ...)` raises, so `MyArr.from_numpy` is never called (and thus the `__array_ufunc__` call is effectively meaningless).

mhvk replied:
Ah, I see. The class would then also have to implement `__array_wrap__` or write `__eq__` better. I guess that just has to be the case, since `np.equal` and `ndarray.__eq__` work differently.

mhvk commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1074,33 +972,100 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+         * so simply fill it with 0 or 1.
          */
-        result = _failed_comparison_workaround(self, other, cmp_op);
+        memset(PyArray_BYTES(res), cmp_op == Py_EQ ? 0 : 1, PyArray_NBYTES(res));
```
I'd tend to just use a standard `fill` function since this is not a critical path at all, but obviously not a big deal.

mhvk commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1074,33 +972,100 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
+    /*
+     * At this point we are locked into comparing (both are arrays) or
+     * something we would have tried to convert.
+     * If we are not in `==` and `!=`, this is an error and we hope that
```
Is this not a change in behaviour for flexible `dtype`? Previously, that would return `NotImplemented` and give a `TypeError`. But I think a clearer error would be a win, so this is really just to check it is intended.

seberg replied:
I think we are fine?  We have strings supported in the ufunc now and the void dtypes are explicitly handled.  There are no further flexible dtypes left, since users effectively cannot create them.

mhvk replied:
Indeed! Which is very nice!

mhvk commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -1622,6 +1622,13 @@ compute_datetime_metadata_greatest_common_divisor(
+ * should promote (two datetimes should have a common datetime), but cannot.
+ * The promotion is thus valid, but fails.
+ * This is important, because `arr == arr` for these cannot reasonably return
+ * all ``False`` values.
```
Does it fail instead? Maybe extend sentence "... values, but should fail instead."?

seberg replied:
Ah, should expand on this... The cast normally fails because it is considered "unsafe".  The old hack is unnecessary because we now find the loop directly in either case.

mhvk commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4811,6 +4811,49 @@ intern_strings(void)
 
+
+/*
+ * Initializes global constants.  At some points these need to be cleaned
```
Agreed that it would be good to clean this up -- this setting of static globals seems better than caching it in many different places.

mhvk commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -1018,12 +1016,29 @@ promote_and_get_ufuncimpl(PyUFuncObject *ufunc,
+     * then we chain it, because InvalidPromotion effectively means that there
+     * is no loop available.  (We failed finding a loop by using promotion.)
+     */
+    if (PyErr_ExceptionMatches(npy_InvalidPromotion)) {
```
`else if`?

mhvk commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -9493,6 +9500,104 @@ class MyAlwaysEqualNew(MyAlwaysEqual):
+@pytest.mark.parametrize(["dt1", "dt2"], [
+        ("M8[ns]", "d"),
+        # Note: timedelta currently promotes (and always did) :(
+        #       so it does not line in here.
```
I thought you are fixing that too?  (also spelling: "line"?)  Or promotes to int?

seberg replied:
I really really want to pull the plug in this as well!  But here I only changed datetime and timedelta, not timedelta and integer promotions (I think we should keep the two changesets distinct)

mhvk commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -9493,6 +9500,104 @@ class MyAlwaysEqualNew(MyAlwaysEqual):
+    # Make sure they broadcast to test result shape, use random values, since
+    # the actual value should be ignored
+    arr1 = np.random.randint(5, size=100).astype(dt1)
+    arr2 = np.random.randint(5, size=99)[:, None].astype(dt2)
```
nitpick: `np.newaxis` is clearer.

mhvk commented on numpy/core/tests/test_regression.py:
```diff
@@ -128,10 +128,7 @@ def test_scalar_compare(self):
         assert_(a[1] == 'auto')
         assert_(a[0] != 'auto')
         b = np.linspace(0, 10, 11)
-        # This should return true for now, but will eventually raise an error:
```
Hmm, comment was incorrect...

seberg committed:
MAINT: Fix pointer cast subclass check and for exceptions

just silences a compiler warning.

mhvk commented on numpy/exceptions.py:
```diff
@@ -195,3 +196,49 @@ def __str__(self):
+
+    Notes
+    -----
+    ``InvalidPromotion`` derives from ``TypeError`` and if it occurs within
```
replace "if it occurs within" with "in"

seberg replied:
rephrasing a bit, the use of parentheses wasn't nice anyway.

mhvk commented on numpy/exceptions.py:
```diff
@@ -195,3 +196,49 @@ def __str__(self):
+    Datetimes and complex numbers are incompatible classes and cannot be
+    promoted:
+
+    >>> np.result_type(np.dtype("M8", "D")
```
Example is scrambled; needs two `np.dtype`?

mhvk commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -9493,6 +9500,104 @@ class MyAlwaysEqualNew(MyAlwaysEqual):
+        arr1 != arr2
+
+    # Basic test with another operation:
+    with pytest.raises(np.core._exceptions._UFuncNoLoopError):
```
Is this exception not exported? Should it?

seberg replied:
I agree, we should, even if just that users should be able to do the same `try: np.equal()` hack that we can.  I would also be happy to keep it a followup, if we can lock in the promotion error here.

mhvk replied:
Agreed, this can be done separately, exposing all of those `_exceptions`

mhvk reviewed (commented):
This looks good! It seems my main comment is about the name of the new exception...

seberg commented on numpy/core/_internal.py:
```diff
@@ -9,6 +9,7 @@
 import sys
 import warnings
 
+from ..exceptions import InvalidPromotion
```
I once heard the opinion that having "error" or "exception" inside an Exception is really redundant, but no strong opions.
I do agree that there is a point in including `DType` in the name though.

mhvk replied:
In principle agreed, but given that essentially all CPython exceptions have "Error" in the name... I think I also prefer having `DType` in the name. So, maybe `DTypePromotionError`?

mhvk commented:
p.s. This fails one astropy test, but it was one that tested the status quo; having this would be an improvement!

seberg committed:
MAINT: Remove pedantic check that exception is an exception

Python does even crash when you get it wrong, so this is unnecessar

seberg commented:
@mroeschke I am curious about pandas.  I am seeing a lot of failures in this test:

<details>

```
    @pytest.mark.filterwarnings("ignore:elementwise comp:DeprecationWarning")
    def test_get_indexer_non_unique_np_nats(self, np_nat_fixture, np_nat_fixture2):
        expected_missing = np.array([], dtype=np.intp)
        # matching-but-not-identical nats
        if is_matching_na(np_nat_fixture, np_nat_fixture2):
            # ensure nats are different objects
            index = Index(
                np.array(
                    ["2021-10-02", np_nat_fixture.copy(), np_nat_fixture2.copy()],
                    dtype=object,
                ),
                dtype=object,
            )
            # pass as index to prevent target from being casted to DatetimeIndex
            indexer, missing = index.get_indexer_non_unique(
                Index([np_nat_fixture], dtype=object)
            )
            expected_indexer = np.array([1, 2], dtype=np.intp)
            tm.assert_numpy_array_equal(indexer, expected_indexer)
            tm.assert_numpy_array_equal(missing, expected_missing)
        # dt64nat vs td64nat
        else:
            index = Index(
                np.array(
                    [
                        "2021-10-02",
                        np_nat_fixture,
                        np_nat_fixture2,
                        np_nat_fixture,
                        np_nat_fixture2,
                    ],
                    dtype=object,
                ),
                dtype=object,
            )
            # pass as index to prevent target from being casted to DatetimeIndex
>           indexer, missing = index.get_indexer_non_unique(
                Index([np_nat_fixture], dtype=object)
            )


>   ???
E   OverflowError: Integer overflow getting a common metadata divisor for NumPy datetime metadata [as] and [m].
```

</details>

Mostly failing due to incompatible time units (NumPy cannot convert between them and now raises an error somewhere deeper in `index.get_indexer_non_unique()`).

Failing for incompatible time units like that seems rather minor on first sight.  We could make whatever comparison is failing here pass if we argue that the values cannot possibly be equal.  (The DeprecationWarning didn't warn of that possibility, but if we are sure of the result it makes sense to change, IMO...)

seberg committed:
DOC,TST: address review comments and improve test coverage

Co-authored-by: Marten van Kerkwijk <mhvk@astro.utoronto.ca>

seberg committed:
MAINT: Rename to DTypePromotionError

seberg committed:
DOC: Add release note for promotion/ and ==/!= expired deprecation

seberg committed:
DOC: No need to repeat "error" after DTypePromotionError and line breaks

seberg commented:
OK, decided to change to ``DTypePromotionError``.  I agree that ``DType`` should be in there and "InvalidPromotion" reads no or barely better than "PromotionError", so there is no reason not to follow the typical patter.

seberg committed:
BUG: Add missing error check

mhvk commented on numpy/core/src/multiarray/datetime.c:
```diff
@@ -1622,6 +1622,15 @@ compute_datetime_metadata_greatest_common_divisor(
+     * The following errors do not currently use "DTypePromotionError".  This
+     * could be replaced (changing the OverflowError, though) with an explicit
+     * promotion error.
+     * This makes sense if we `arr_dt1 == arr_dt2` can reasonably return
```
"if we" -> "since"?

I have a bit of trouble following the comment. Maybe easier to write it as
```
It might make sense to replace the following errors with "DTypePromotionError",
since we are here unable to convert one dtype to another.

But one would have to think of whether one then wants `arr_dt1 == arr_dt2` to return
an array of all `False`.
```
Aside: I don't think it should return an array of all `False` for either case, since it is just that
the conversion is difficult, not that it is not possible in principle. And some of the times
might well actually be equal (e.g., if they are 0?). So, I think an error is actually better.

seberg replied:
Ah, good point about 0, I think that is a killer argument (until we wish to change the `==` implementation, but even then I think the argument makes sense).

mhvk commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -1291,26 +1298,29 @@ def test_structured_comparisons_with_promotion(self):
 
-    def test_void_comparison_failures(self):
+    @pytest.mark.parametrize("op", [
+            operator.eq, lambda x, y: operator.eq(y, x),
```
Not quite sure why the `lambda`s are also tested...

seberg replied:
They test the reverse direction.

mhvk replied:
Ah, missed that...

mhvk reviewed (commented):
Looks great. Only trivial comments.

mroeschke commented:
> @mroeschke I am curious about pandas. I am seeing a lot of failures in this test:

> Mostly failing due to incompatible time units (NumPy cannot convert between them and now raises an error somewhere deeper in index.get_indexer_non_unique()).

From a glance given the exception, integer overflow due to time unit conversion, that exception reason generally matches what pandas will do for supporting other resolutions other than nanoseconds (e.g. `datetime64[us]`) so I want to say it's reasonable to this to fail.

seberg committed:
DOC: Tweak subclass in richcompare comments based on review

seberg committed:
DOC: Slightly expand ==/!= release note with examples

seberg committed:
DOC: try to clarify datetime not using DTypePromotionError comment

mattip commented on doc/release/upcoming_changes/22707.expired.rst:
```diff
@@ -0,0 +1,13 @@
+  fundamentally not comparable (e.g. have different dtypes).  An example
+  is ``np.array(["a"]) == np.array([1])`.
+
+For a long time these gave ``DeprecationWarning`` or ``FutureWarning``.
+This mimics the Python behavior of returning ``False`` and ``True``
+when comparing incompatible types like ``"a" == 1`` and ``"a" != 1``.
```
```suggestion
This mimics the Python behavior of returning ``False`` and ``True``
when comparing incompatible types like ``"a" == 1`` and ``"a" != 1``.
For a long time these gave ``DeprecationWarning`` or ``FutureWarning``.
```

mattip commented on doc/release/upcoming_changes/22707.improvement.rst:
```diff
@@ -0,0 +1,9 @@
+    np.result_type("M8[s]", np.complex128)
+
+Errors that occur while trying to find the correct promotion may raise a
+different error.
```
This sentence is incomplete. Either remove it, or give more detail: what other error class can be emitted?

seberg replied:
Yeah, this is mainly the weirder datetime unit conversions that fail.  But maybe that is all obscure enough that removing the sentence is best, since I don't want to give the example.
(Will remove it)

mattip commented on numpy/__init__.py:
```diff
@@ -107,6 +107,8 @@
 
 from ._globals import _NoValue, _CopyMode
+from . import exceptions
+# Note that the following names are imported explicitly for backcompat:
```
Do we really have to export Python errors/warnings in the NumPy top-level namespace? If so those should be deprecated.

seberg replied:
I agree we should follow-up with deprecating or at least hiding them from the main namespace.   We could move the "namespace", I held off with that in the first PR, but the only thing that might notice is unpickling in an older version, and that is probably not worth the trouble for exceptions.

mattip commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1074,33 +972,99 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
-    if (result == NULL) {
+
+    /*
+     * At this point we are locked into comparing (both are arrays) or
+     * something we would have tried to convert.
```
I don't understand the comment above: is the closing parenthesis in the wrong place? If we tried to convert it, did we fail? I would prefer it said this, but maybe it is not true.
```suggestion
     * At this point we comparing two ndarrays.
```

mattip reviewed (commented):
A few suggestions and questions. Martin already did most of the review.

seberg commented on numpy/core/src/umath/dispatching.c:
```diff
@@ -43,6 +43,7 @@
 #include <convert_datatype.h>
 
 #include "numpy/ndarraytypes.h"
+#include "numpy/npy_3kcompat.h"
```
Yah... its for the error chaining, which uses an underscored function from CPython otherwise.

seberg committed:
Apply suggestions from code review

Co-authored-by: Matti Picus <matti.picus@gmail.com>

seberg committed:
DOC: Remove last note from release note as per suggestion

It is a bit too obscure, but explaining that it happens for datetime
unit conversions seems more detailed then useful.

Updated to include a last non-indent line, which the release notes
machinery doesn't like otherwise...

seberg committed:
DOC: Tweak comparison code comment to be a bit more clear

mattip reviewed (approved):
LGTM. Just one typo. I will push a fix, feel free to merge once CI passes

mattip committed:
DOC: typo

seberg commented:
Wow, that was a quick turnaround on this, thanks Matti and Marten!  Fingers crossed that downstream doesn't mind the FutureWarning's suddenly being fixed :).

If anyone has an opinion about the error naming or anything else, I am happy to follow up of course!

h-vetinari commented:
I realise it's probably too late but given that this fixes a pretty long-standing issue (or at least annoyance), should this be backported to 1.24?

charris commented:
> this fixes a pretty long-standing issue

That is an argument for _not_ backporting :)  If it only fixed one simple issue, maybe, but I think this is more expansive and should probably sit in the development branch for a while.

neutrinoceros commented:
I'd like to add another argument for _not_ backporting (as is, at least): this patch caused a regression downstream (see https://github.com/numpy/numpy/issues/22744)

mhvk commented:
Indeed, it also fails one astropy test (see https://github.com/numpy/numpy/pull/22707#issuecomment-1334147730), though it is one that we are very happy to fix. But it means one definitely should *not* backport - much better if numpy version checks can be by minor number only, not micro.

h-vetinari commented:
> much better if numpy version checks can be by minor number only, not micro.

1.24.0 has not been released yet, so there'd be no need to drop to micro level.

I see the argument that it should sit in main for a while, though as pointed out the "breakage" often a positive thing due to removing the workarounds that were necessary for this issue.

I still find it a pity that we'll have to wait another ~6month for this. Had this PR been merged 2 weeks ago, its inclusion in 1.24.0 wouldn't be controversial.

But oh well, those 6 months will pass as well 🤷

mhvk commented:
Hmm, I had not realized numpy 1.24 was not out. Still, probably best to give people time to fix bugs; astropy and others test against -dev just for that purpose, and a bit of extra time is good.
